### ISSUE TITLE
Showing output wavefield with respect to undulated, physical mesh, and also confirming representation of input model

### ISSUE THREAD

**Issue by axelwang** [AxiSEMunity/AxiSEM3D#31]
These questions concern a 2D cross-section of a 3D model.

For my specific use case of AxiSEM3D, it is very important to show how the wavefield interacts with specific 2D structures. For example, I would like to plot snapshots like the following example (Figure 1; Furumura et al., 2014) that shows how seismic waves pass through a region with undulating Moho (and surface) which creates varying crustal thickness and most prominently, a pinched crust under the Sea of Japan.  

![Figure 1](https://user-images.githubusercontent.com/45047232/189274674-0d24d857-eef8-4712-8b31-2ec84a5f8350.png)

To do this, I looked into example 3 (the salt body example) that has element output as well as how AxiSEM3D is handling 2D variations (with particle relabeling transformation, or, mesh stretching). The important idea is captured in the following figure from Al-Attar and Crawford, 2016. 
![Figure 2](https://user-images.githubusercontent.com/45047232/189274912-6cba3d6c-e9db-4409-a7a1-48ed42ce6b61.png)

 I will spare my helpers from reading me explaining this in detail but I would like to check my understanding and perhaps help others on what the code is actually doing: 

- Read in a 3D model (for simplicity we can assume we read in only a slice of that 3D model, i.e. the black part of the left figure)
- Find $\tau(r)$ (Eq. (5) of Leng et al, 2019) and use that to change the irregular mesh (left) to a regular one (right). This is particle relabeling transformation and it is also applied to functions like the source that was initially defined on the irregular mesh (left). $\tau(r)$ is not easily accessible. 
- Simulation is performed on the regular mesh (right) and we obtain the distorted wavefield shown on the right. 
- Inverse particle relabeling transformation is performed on the distorted wavefield (on the right) and we obtain the real, physical wavefield shown on the left. BUT THE MESH IS NOT TRANSFOMRED BACK.

I am not entirely certain if the above is accurate, especially for the last point, which is motivated by what I saw from the output of example 3. 

![Figure 3](https://user-images.githubusercontent.com/45047232/189275136-77f40b21-2bde-4391-9f1a-ff574522fb57.png)

Here we can see a wavefield snapshot plotted for example 3, using its included post_processing.ipynb script. Each point (intersection of horizontal and vertical lines) is an output GLL point, and the output option here is set to `[0, 2, 4]: 4 vertices + 4 edge centers + 1 element center`, so four squares in the above mesh correspond to 1 quad element. The top of the mesh corresponds to the ocean floor, which is undulated. For example, according to the 3D geometric model, at the first frame (y=0), the ocean floor looks like this:
![Figure 4](https://user-images.githubusercontent.com/45047232/189275221-1dc4ee12-0c54-47da-b439-bcc210783a58.png)

and we can see the max range of undulation (from x=0 to 5000 m only, the range actually used in the model) is about 40 m. The undulation range is set to `min_max: [0., 1000.]` for this geometric model, which means from 6370.8km up to 6371 km (not shown on the above mesh because this is a water layer), and down to 6369.8 km (shown in the above mesh), all horizontal edges should be deformed/undulated. This region is about the top 3rd of the mesh shown above. However, as can be seen in the plot (all lines are straight) as well as confirmed by me looking at the coordinates of these GLL points, these coordinates form a regular mesh (similar to right panel of Al-Attar's plot) instead of the deformed mesh (similar to left panel of Al-Attar's plot), and those deformed (i.e. the physical one) GLL coordinates are no where to be found in the output, nor is the $\tau(r)$ function that could be used to apply the transformation.  This is why I am saying it seems like the inverse particle relabelling transformation is not being performed on the mesh. 

I am fairly certain that the wavefield output by AxiSEM3D is not the distorted one (something like the right panel of Al-Alttar's plot) but it should be the real, physical wavefield (something like the left panel of Al-Attar's plot). **However, I think such a wavefield should be plotted on the physical mesh instead of the regular mesh?** In fact, plotting on regular mesh will distort this real wavefield solely due to plotting.  

I am scratching my head on how to fix this. For a start, one needs to change the source code to also perform inverse particle relabelling transformation (this inverse transformation should already exists and is applied to the wavefield) to the mesh as well. This way, we can obtain the GLL points coordinates in the physical mesh (or simply output $\tau(r)$ so the inverse transformation on the mesh can be done as post-processing). But now there is also the question about how to connect these GLL points in the physical mesh. I think the shape function used for meshing in AxiSEM3D is only order 1, meaning all element edges are straight lines in the regular mesh. But is this still true for the stretched/physical mesh?  In addition, does $\tau(r)$ really need to be $\tau(r,\theta)$? I am asking this because of Figure 2 in Leng et al., 2019, also shown below. We can see that all points on the same horizontal line as P in (a) should have the same $r$, but different $\theta$, and in (b) we see that those points next to P are moved/stretched by different amounts than P, hence it does seem $\tau$ is varying for different $\theta$. Also from this figure, it seems that in the physical mesh, the element edges indeed become curved, but I am guessing each GLL point within an element is still connected with straight lines?

![Figure 4](https://user-images.githubusercontent.com/45047232/189465942-45ce8974-7dce-4a22-8f1c-156f66cd4ddf.png)

In summary, what I need help with is to find a way to output GLL coordinates in the physical mesh and figure out how to connect them. Once this is done, then the wavefield can be plotted on this set of GLL points and it should be free of visual distortion (the wavefield is already being output for each output GLL points, it's just that the coordinates of these points correspond to a regular mesh). This will also allow me to do what I set out to do, i.e. to plot something like the first figure above.

Thank you very much for reading through this and I hope to hear what do you all think. Please correct me if I am misunderstanding anything. 

-Axel

**Comment by kuangdai** [comment]
1. The displacement output is physical and unique, regardless of which grid (flat or deformed) you plot it on. For example, in the figure from Al-Attar and Crawford, 2016, the colors in the left and the right plots arue exactly the same -- that's the point of the paper. It is important to understand this.
2. Understanding 1,  you will see that plotting displacement on a flat or deformed grid is just a choice of visualization -- nothing related to the wave physics. For example, at a continental scale, the surface undulation (10 km) can be much smaller than the computational model (1000 km), and we usually exaggerate the mesh deformation a lot. 
3. How you do that? The coordinates of the deformed grid is simply (x, y+ A τ), where (x, y) are the flat coordinates, τ the undulation, and A your choice for best visual effect. When A=1, you get the actual shape of the model used for computation. Now the example has created a vtk file to make the animations, so (x, y) are already there in the vtk. You need to add a nodal variable τ (computed based on your input), and send it to the vtk. Finally, you can create y'=y+Aτ in Paraview and switch the y-coordinate varible from y to y' to deform the mesh. Note that I am suggesting you to compute y' in Paraview instead of creating y' directly in the vtk file because you can choose A dynamically to see which one looks the best. As a community code for computation, AxiSEM3D is only reasonably versatile for wavefield output, not as powerful as commerical FE software. You need to add some lines in the notebook to add τ to the vtk files.

**Comment by axelwang** [comment]
Hey @kuangdai. Thank you so much for the explanations! I am still wondering though, how do I exactly add $\tau$ to the (x,y) coordinates? Ideally, since this is already computed by the code, I would like to just ouput exactly what is used by the code. Currently, I am looking at issue #18 where at the end you made a plot of the GLL grid points with undulations on it (though for that example, the undulations are not that obvious). I think `geodesy::sz2rtheta(getPointSZ(), false)` is the key but I don't know how to run only an individual function from a large set of C++ codes. So if you can let me know a bit more about how you made the last plot in #18, it would be very appreciated!!
There is also the issue of how do I define the boundaries of the elements once the grid is deformed. Does each element still have straight edges? If so, then it will be easy to just also connect all GLL points within each element by straight lines as well. I am thinking, for example, if I just choose one displacement value per element (say the center GLL point), then the coloring corresponding to this element should follow the exact shape of the deformed element, but not just a square as standard plotting algorithms will do, otherwise the wavefield will not follow the undulating boundaries.

**Comment by axelwang** [comment]
Hey @kuangdai. I am sure this is probably not the best/most elegant way of doing it, but it seems I have found a way to access those undulation data and I just want to confirm with you if this makes sense. 
In `Quad::computeDt`, you have `const eigen::DMatXN &dZ = mUndulation->getElemental();` under comment _//compute coords on slices_. I output those `dZ` values and seems like they contain the amount of undulations on all GLL points for all elements and for all azimuthal slices. 
Furthermore, a bit further down this same function, under comment _//both 1D and 3D undulation_, we have the `sz` variable which contains the physical (regular mesh + undulation) coordinates  for all GLL points for all elements and this is looped over all azimuthal slices. 
Again, this is probably an awkward way to access those but these sections of codes is what made the most sense to me regarding what I am looking for, so far...

I haven't managed to make a plot yet, which brings me to my next, and hopefully final question, regarding this topic: Do I just connect those GLL points with straight lines? 

If so, will this make kinks on element edges? I am thinking yes but those probably are hardly kinks but rather it's going to be pretty smooth.  I am worried about connecting them because this defines a region on which I shall paint some color to represent some field. 
Say if I want to represent some field on the deformed mesh (this could be the wavefield or the velocity or density etc.), and I will just use the value on the central GLL point for each element to represent the whole element. When I have a spherical mesh such that each element is not an upright rectangle in the cartesian coordinates (most plotting algorithm colors an upright rectangle region), don't I need to color each element following its exact shape? I have managed to do this (see attached figure; the mesh is the same as the one for example 2, the field plotted is VPH),  
![regional_mesh__prem_ani__50s](https://user-images.githubusercontent.com/45047232/190968296-3b8c843f-1ae0-40e0-a70c-69f359da2357.png)
it seems like with the GLL points now added to this (and if they are connected by straight lines), now each element is actually 16 smaller sub-elements. This algorithm is pretty slow so I am wondering what is a standard way for plotting fields are meshes like this, even without the undulations?

**Comment by axelwang** [comment]
> Which pair in () correspond to which point in the regularGLL.txt? This I really don't have a clue but I have tried pretty much all possible ways... The most sensible thing I can think of is that, note there is pretty much always (though sometimes not true) a column of 0's for all orders within each element, now that has to correspond to an edge that is not deforming. For the first row of elements (i.e. 1450-1499), they have to connect with the below elements that are not deforming, otherwise the mesh becomes disconnected at the transition to deforming regions, so for these elements that column of 0's have to correspond to the 5 GLL points on the bottom edge.

I figured out the correspondence here by outputting some data in `Quad::computeDt` under comment `// both 1D and 3D undulation`. I attach this output here as `testOutputs.txt`. I am almost certain that I got the right Fourier coefficients and the way I am adding them up is correct, as my code above can produce the same dZ's for $nr=0$ and $nr=50$, which seems to be corresponding to the same cardinal point, i.e. $\phi=0$ and $\phi= 2 \pi$?  However, in `testOutputs.txt`, what is baffling me is that it seems for $nr=0$, the $dZ$ is with respect to the original coordinates `szRef`, which is what I output above in `regularGLL.txt`, however, for subsequent $nr$'s, the reference is not `szRef` anymore, but  with respect to the `sz` from the previous $nr$? 

> Elements included are only those within the undulation range, but they are cycled twice (i.e. Element 1450-1999 are cycled twice, with distinct and non-zero "fourier coeffficients".) This makes sense because the 3D model and spectral model are built twice in the code (for MPI purpose?). For now, I have only considered the second time when the "coefficients" are output for those elements, i.e., I discard half of the output in fCoeffs.txt (as can be seen in my ipynb). Since I don't know why the models are built twice, I am wondering is what I am doing correct?

From the `undulation_range` parameter for example 3, the elements that are within this range are actually 1500-1999, rather than 1450-1999, but this is what `Undulation::finished3D` is outputting so there is probably some subtle reason here. The fact that 1500-1999 is undulating is reflected in `testOutputs.txt`, as only starting from element 1500 we start to see none-zero dZ's, so I am sure the code is doing what it is supposed to do. Nonetheless, for constructing the physical mesh, it seems I need to simply ignore the output Fourier coefficients for elements 1450-1499. This actually helped a lot because I didn't have to align the column of 0's to stitch the undulated mesh and regular mesh together, and yet I still don't have a gap in between them. The new mesh looks quite good now with the correct correspondence and undulating element range:


![newplot (16)](https://user-images.githubusercontent.com/45047232/192218943-3694f085-3241-474e-aca1-0de913ea5cfb.png)

I still have 3 questions noe:

- Reiterating my first point above. How do I deal with other $\phi$'s. It seems the reference is not always the undeformed grid.
- The above plot is supposed to be for $\phi=0$. The `GEO` model for this example has an azimuthal `y` value with size 673, I can't seem to find a frame (i.e. a `y`) that actually corresponds to what I have on the mesh above. The discontinuity is supposed to be the bottom edge of the second row of elements. 
- With the plotting above, it seems shared GLL points on element edges now no longer align perfectly (when these were simply repeated values in the regular mesh file). The vertical offsets now are about a few meters. See a zoom in of the above mesh below. I am guessing this will pose a serious problem when the wavefield is output on those points? Because now you essentially have two locations for one value of the wavefield... (unless the above mesh is still incorrect and those shared GLL points between elements should still remain simply as repeated identical values). But I am guessing this is probably fine if you just plot the wavefield value on the central GLL point for each element? Here is an example of the points mismatch I am seeing:
![newplot (17)](https://user-images.githubusercontent.com/45047232/192221656-86cad703-6251-4d43-bb05-d6b65494c76d.png)

**Comment by kuangdai** [comment]
Thanks. 

The error in your plot is probably because you didn't use the Fourier coefficient correctly.

To get what you want:
1. Replace `Undulation.cpp` with the attached one, recompile and run (note: you can change `max_num_time_steps` in `inparam.advanced.yaml` to 1 to stop the similation fast). 
2. Once you get the output folder, copy `undulation.ipynb` to the simulation folder. Running it, you will get a vtk of deformed mesh. You can change the phi value to relocate the slice.
3. Drag the vtk to Paraview to get below:

![und](https://user-images.githubusercontent.com/9271974/193110673-0cfbfc9d-b718-4490-bc4a-e73e5ecc9184.png)

[undulation.zip](https://github.com/AxiSEMunity/AxiSEM3D/files/9678034/undulation.zip)

**Comment by axelwang** [comment]
Hey @kuangdai. I haven't looked closely in your codes yet, but just very quickly, how does the mesh you show correspond to the input 3D GEO model for the ocean floor? 
I think the simulation domain is only the right half (0-5000 m) of the input model right? I am not seeing which frame (the number of y values in the GEO model), precise or approximate, shows a downward ramp toward the right of the figure, i.e. larger x? 
I attach a script here so you can quickly look at the GEO model. 
[plot_2DSlice_models_example3GEO.ipynb.zip](https://github.com/AxiSEMunity/AxiSEM3D/files/9678643/plot_2DSlice_models_example3GEO.ipynb.zip)

**Comment by kuangdai** [comment]
Debugging your code can take much longer. You can use my code for the same purpose. Attached is an updated notebook for 3D plots like above.

 
[undulation.ipynb.zip](https://github.com/AxiSEMunity/AxiSEM3D/files/9678803/undulation.ipynb.zip)

**Comment by axelwang** [comment]
Of course. Thank you so much. Is my understanding correct that one of the 670 ish $y$ values in the GEO model also corresponds to one such slice like you show above? If yes, then the easiest thing to check is $y=0$ which should be mapped to $\phi=0$? Though this is not matching (i.e. the undulated mesh for the seafloor and the surface from GEO model are pretty different.)

**Comment by kuangdai** [comment]
What is GEO model? SEG_C3_UNDULATION.nc in the  input file?

**Comment by axelwang** [comment]
Sorry. I meant the 3D geometric model. Yes that's the file name.

**Comment by axelwang** [comment]
One of the things I am checking to see if I did the mesh deforming correctly is to plot the $x$ vs. $depth$ at different $y$ values and compare with the deformed surface (just the bottom edge of the second row of elements from top) from the mesh. The problem is I fail to see correspondence.  

Can you confirm if my understanding is correct that these 676 $y$ values in the model is sampled over 360 degrees, so about 0.53 degree per $y$ frame. This means the mesh at $\phi=0$ should correspond to the model at $y=0$ and  the mesh at $\phi=90$ should correspond to a $y$ value of about 90/0.53 $\approx$ 170? 


Below is what the model shows at $y$= 169
![newplot (21)](https://user-images.githubusercontent.com/45047232/193137820-0367c51b-1cdd-4117-ac3f-3ae6b69b730d.png)

**Comment by kuangdai** [comment]
Ok. I understand your question now.

The relation between $(x,y)$ and $\phi$ is Cartesian-polar:
$x=s \cos(\phi)$
$y=s \sin(\phi)$
where $s$ is distance from the earthquake or the centre $x=0, y=0$.

Therefore,
* $\phi=0^\circ$ is the slice $y=0$, $0< x<5000$.
* $\phi=180^\circ$ is the slice $y=0$, $-5000< x<0$.

So $\phi=0^\circ$ and $\phi=180^\circ$, together they make the plane $y=0$. 


And $\phi=90^\circ$ is the slice $x=0$, $0< y<5000$,  so to show  that you need to plot YZ and use X in your slider. 

To verify:

This is your plot at $y=0$ in the first post:
![image](https://user-images.githubusercontent.com/9271974/193147376-cb37a511-49d1-4306-bcd9-d86dfbfb89e0.png)

This is the deformed mesh generated from the ipynb I just sent, including two slices, $\phi=0^\circ$ on the right, and $\phi=180^\circ$ on the left:
![0+180](https://user-images.githubusercontent.com/9271974/193148592-ec2db2c1-5616-4ccd-9547-f264cec93a53.png)

They look right.

**Comment by axelwang** [comment]
Hey @kuangdai. Thank you so much for the explanations. It is very helpful! 

I also found the bug in my own code, which was just an additional flipping up-side-down for the Fourier coefficients ordering. 

I also made another mistake, and that's regarding plotting the physical model:
![193147376-cb37a511-49d1-4306-bcd9-d86dfbfb89e0](https://user-images.githubusercontent.com/45047232/193438603-2b0735fe-d2b3-4dda-9f75-659edd532b1c.png)

This is actually NOT $y=0$, but rather it is just the first value of $y$, which is $-5062.5$. This explains why the meshes you showed above actually don't quite match the fake $y=0$ plot I made. Most notably, in the red mesh, $\phi=0$, you can clearly see a subsidence whereas this is not there in the fake $y=0$ plot. 

The depth for ocean floor is given in a 676x676 cartesian grid in the input model, so the only $\phi$ values that are naturally represented are the diagonals, i.e. $45-225^\circ$ and $135-315^\circ$. $\phi=0^\circ$  and $\phi=180^\circ$ is actually not naturally represented because there is no $y=0$ there, the closest is $y=\pm7.5$.  I plotted all $x$ at $y=-7.5$ from the input model, and also the ocean floor on the mesh. Here we see they match visually perfectly:
![newplot (20)](https://user-images.githubusercontent.com/45047232/193438998-8eab8ea0-1206-4cf7-a19a-f68f61d137e9.png)
(you can see that subsidence around $x=800 - 2000$)
The same is done for $\phi=90^\circ$ and $\phi=270^\circ$:
![newplot (15)](https://user-images.githubusercontent.com/45047232/193440808-a67d6ddd-0d55-4ef1-a17c-bb964d08e95e.png)


But I am having trouble doing this for $45-225^\circ$:

![newplot (20) copy](https://user-images.githubusercontent.com/45047232/193439240-2a6c1c13-86de-4716-b5a8-5c89a5dcce09.png)

as well as for $135-315^\circ$:
![newplot (21)](https://user-images.githubusercontent.com/45047232/193440831-f7c7acbf-0991-4302-a4c8-516b867d18d8.png)

These diagonal planes are the only 4 planes that are exactly represented by the model (i.e. without having to interpolate the input model or find something close, such as what I did for the axial planes) yet I can't seem to find their exact representation by the mesh... I am sure these should be exactly represented and this is probably again some small ordering error in my code, but for now I am stuck on debugging these. Are these digonal planes not represented by the diagonal components in the $depth$ variable in `SEG_C3_UNDULATION.nc`?

**Comment by axelwang** [comment]
Ah. I realized what's going on as soon as I posted.
Even though the number of points on the diagonal slices are the same as $x$, their sampling interval is $\sqrt{2}$ the $x$ interval, hence even for these slices that I thought you don't need to interpolate, you actually have to interpolate... Once you do that, everything works out:
![newplot (28)](https://user-images.githubusercontent.com/45047232/193442192-802ffd48-fea4-439c-a9b5-ba23600cc74f.png)


I think this can be the end of this long thread. Thanks to @kuangdai  for helping tremendously!
